feat: add hCaptcha registration form action

warioishere opened the pull request:
adds hcaptcha as requested in https://github.com/keycloak/keycloak/issues/41057

- generalize the registration captcha base so providers can supply their response field names, template metadata, and scripts
- add an hCaptcha form action with service registration, theme updates, and documentation for configuring the provider
- extend admin provider tests to expect the hCaptcha action in addition to the existing reCAPTCHA entries

mabartos commented on docs/documentation/server_admin/topics/users/proc-enabling-recaptcha.adoc:
```diff
@@ -6,8 +6,8 @@
-To safeguard registration against bots, {project_name} has integration with Google reCAPTCHA (see <<procedure_recaptcha>>) and reCAPTCHA Enterprise (see <<procedure_recaptcha_enterprise>>).
-The default theme (`register.ftl`) supports both v2 (visible, checkbox-based) and v3 (score-based, invisible) reCAPTCHA (see https://docs.cloud.google.com/recaptcha/docs/choose-key-type[Choose the appropriate reCAPTCHA key type]).
+To safeguard registration against bots, {project_name} has integration with Google reCAPTCHA (see <<procedure_recaptcha>>), hCaptcha (see <<procedure_hcaptcha>>), and reCAPTCHA Enterprise (see <<procedure_recaptcha_enterprise>>).
+The default theme (`register.ftl`) supports both v2 (visible, checkbox-based) and v3 (score-based, invisible) reCAPTCHA (see https://docs.google.com/recaptcha/docs/choose-key-type[Choose the appropriate reCAPTCHA key type]).
```
```suggestion
The default theme (`register.ftl`) supports both v2 (visible, checkbox-based) and v3 (score-based, invisible) reCAPTCHA (see https://docs.cloud.google.com/recaptcha/docs/choose-key-type[Choose the appropriate reCAPTCHA key type]).
```

The new link does not work.

mabartos commented on testsuite/integration-arquillian/servers/auth-server/services/testsuite-providers/src/main/resources/theme/address/login/register.ftl:
```diff
@@ -107,10 +107,15 @@
+                        <div class="h-captcha" data-size="compact" data-sitekey="${recaptchaSiteKey}"></div>
+                    <#else>
+                        <div class="g-recaptcha" data-size="compact" data-sitekey="${recaptchaSiteKey}"></div>
+                    </#if>
```
Would it be possible to generalize it that this register page does not care what CAPTCHA is used? For the class, you could set the form attribute as well.

mabartos commented on themes/src/main/resources/theme/keycloak.v2/login/register.ftl:
```diff
@@ -41,10 +46,14 @@
+                                data-sitekey="${recaptchaSiteKey}" data-callback="onSubmitRecaptcha"<#if recaptchaAction?has_content> data-action="${recaptchaAction}"</#if> type="submit" id="kc-submit">
+                            ${msg("doRegister")}
+                        </button>
+                    </#if>
```
The same as above.

mabartos commented on services/src/main/java/org/keycloak/authentication/forms/RegistrationPage.java:
```diff
@@ -52,7 +52,7 @@ public class RegistrationPage implements FormAuthenticator, FormAuthenticatorFac
     public static final String FIELD_LAST_NAME = "lastName";
     public static final String FIELD_FIRST_NAME = "firstName";
-    public static final String FIELD_RECAPTCHA_RESPONSE = "g-recaptcha-response";
+    public static final String FIELD_RECAPTCHA_RESPONSE = AbstractRegistrationRecaptcha.G_RECAPTCHA_RESPONSE;
```
Is it used somewhere?

mabartos reviewed (changes requested):
@warioishere Nice work! Thanks for the PR. I have added some comments on what could be improved. 

You should also create a release note in the 26_5_0 file to advertise a little bit the new CAPTCHA implementation. 

You should also sign the commit and potentially add the Closes message there. 

Something like:
```
git commit -m "feat: add hCaptcha registration form action" -m "Closes #41057" -s
```

And do not include merge commits here. It should be only one commit - you can leverage the `git rebase`. 

Thanks!

PS: We could do more improvements with CAPTCHA as including it to login pages, using invisible variants, etc. But it's out of scope of this PR. It can be done in follow-up tasks.

stianst reviewed (commented):
We really need to figure out a way to automate testing around this. Lack of tests for reCaptcha has been an issue several times, at least for that one we can easily try it ourselves.

Looks like hCaptcha has a freebie version; that may be sufficient for having some nightly tests or something like that.

warioishere committed:
feat: add hCaptcha registration form action

Closes #41057

Improvements based on code review:
- Generalize documentation title from "Enabling reCAPTCHA" to "Enabling CAPTCHA"
- Fix broken documentation link
- Refactor form attributes to use generic captcha* attributes
- Simplify FTL templates to not hardcode CAPTCHA provider checks
- Maintain backwards compatibility with legacy recaptcha* attributes

Signed-off-by: wario_is_here <mario.hofmann@yourdevice.ch>

warioishere commented:
Thanks for the feedback! I've made all the requested improvements:

## ✅ Administrative Changes
- [x] Added release notes to `26_5_0.adoc` advertising the new hCaptcha support
- [x] Signed the commit with `-s` flag
- [x] Added `Closes #41057` reference
- [x] Cleaned up branch history - removed merge commits, now just one clean commit

## ✅ Documentation Improvements
- [x] Generalized title: "Enabling reCAPTCHA" → "Enabling CAPTCHA" (more inclusive)
- [x] Fixed broken documentation link: `https://docs.google.com/recaptcha/...`

## ✅ Code Refactoring - Provider-Agnostic Architecture
The key improvement is that **templates and forms no longer care which CAPTCHA provider is used**:

### Java Changes
- **AbstractRegistrationRecaptcha**: 
  - Added `setCaptchaAttributes()` method to set generic `captcha*` attributes
  - Added `setRecaptchaLegacyAttributes()` for backwards compatibility with custom themes
  - Added `usesCaptchaCallback()` to indicate if callback-based submission is needed
  - Added abstract `getCaptchaCssClass()` for provider-specific CSS class names

- **RegistrationRecaptcha** & **RegistrationRecaptchaEnterprise**: 
  - Implement `getCaptchaCssClass()` returning "g-recaptcha"
  - Implement `usesCaptchaCallback()` returning `true` (they use callbacks)
  - Set legacy `recaptcha*` attributes for backwards compatibility

- **RegistrationHCaptcha**:
  - Implements `getCaptchaCssClass()` returning "h-captcha"
  - Base implementation of `usesCaptchaCallback()` returns `false` (no callback needed)

### Template Changes
All register.ftl templates now use **generic attributes instead of provider ID checks**:
- Before: Hardcoded checks like `if captchaProviderId == "registration-hcaptcha-action"`
- After: Uses `${captchaCssClass!}`, `${captchaUseCallback!false}`, and generic `captcha*` attributes

Templates are now **completely provider-agnostic** - new CAPTCHA providers can be added without modifying any templates!

## Backwards Compatibility
Custom themes that rely on legacy `recaptcha*` attributes will continue to work. These attributes include TODO comments for future deprecation.

warioishere commented:
so its been a while since you reviewed the PR, is it really in mind of the keycloak dev team to only support a datasucking propietary captcha service?

@stianst
